core/station/impl/src/models/named_rule.rs

```diff
@@ -168,6 +197,31 @@ fn validate_description(description: &Option<String>) -> ModelValidatorResult<Na
     Ok(())
 }
 
+fn validate_policy_compatibility(
+    id: &NamedRuleId,
+    rule: &RequestPolicyRule,
+) -> ModelValidatorResult<NamedRuleError> {
+    let policies = REQUEST_POLICY_REPOSITORY.list();
+    let mut referencing_named_rules = find_all_named_rules_referencing_named_rule(id);
+
+    referencing_named_rules.push(*id);
+
+    for policy in policies {
+        for referencing_named_rule in referencing_named_rules.iter() {
+            if policy.rule.has_named_rule_id(referencing_named_rule) {
+                validate_rule_for_specifier(rule, &policy.specifier).map_err(|e| {
+                    NamedRuleError::IncompatibleWithLinkedPolicy {
+                        policy_id: Uuid::from_bytes(policy.id).hyphenated().to_string(),
+                        error: e.to_string(),
+                    }
+                })?;
+            }
+        }
```

Comment:
```suggestion
                validate_rule_for_specifier_with_substitution(policy.rule, &policy.specifier, id, rule).map_err(|e| {
                    NamedRuleError::IncompatibleWithLinkedPolicy {
                        policy_id: Uuid::from_bytes(policy.id).hyphenated().to_string(),
                        error: e.to_string(),
                    }
                })?;
```
where the function `validate_rule_for_specifier_with_substitution` performs a traversal of `RequestPolicyRule` just like `validate_rule_for_specifier`, but whenever it encounters the named rule with `id`, then it continues the traversal on `rule` instead of the current child. This terminates assuming there are no (indirect) circular references (which we should ensure separately).

I pushed a [counterexample](https://github.com/dfinity/orbit/pull/570/commits/430bd790cb6e040950aaa48b7a74ee800d0ab839) why the current algorithm is not correct.

Reply:
I'd still simplify:
```
/// Validates that the named rule is compatible with the policies that reference it.
/// It traverses all linked policy rules recursively and assumes no circular references.
fn validate_policy_compatibility(
    id: &NamedRuleId,
    rule: &RequestPolicyRule,
) -> ModelValidatorResult<NamedRuleError> {
    let policies = REQUEST_POLICY_REPOSITORY.list();
    for policy in policies {
                validate_rule_for_specifier(&policy.rule, &policy.specifier, &[(id, rule)])
                    .map_err(|e| NamedRuleError::IncompatibleWithLinkedPolicy {
                        policy_id: Uuid::from_bytes(policy.id).hyphenated().to_string(),
                        error: e.to_string(),
                    })?;
    }

    Ok(())
}
```
This way, we only traverse all policy rules once (better worst-case runtime) and the algorithm is simpler (more robust, less changes for logical bugs).

Reply:
Changed as discussed.

Reply:
I dont think that would be good UX. The user would need to understand the nuances of the policy/named_rule/rule system in order to get past it. Let's discuss with Kepler next week?

Reply:
> I dont think that would be good UX. The user would need to understand the nuances of the policy/named_rule/rule system in order to get past it.

I agree, but now we risk having invalid stuff in the upgraded station and potentially getting "stuck" if the validation is on a critical path.

> Let's discuss with Kepler next week?

Indeed.

Reply:
To get stuck what would need to happen? You can always edit a policy rule directly, which should not involve validation of something else. That way you should be able to fix all invalid policies. Then you can fix the named rules if any of them are invalid, since no policies will use them. Can you think of an unrecoverable scenario?

Reply:
E.g., what if there's no valid policy according to the new rules: will the user still be able to execute a request? I guess there are two options:
- fail in post-upgrade if there's any violation of the new rules;
- add end-to-end tests that a violation of the new rules can be recovered after an upgrade.